sim/common/wotlk/enchant_effects.go

```diff
@@ -162,38 +162,31 @@ func init() {
 		character.PseudoStats.ThreatMultiplier *= 0.98
 	})
 
-	core.NewEnchantEffect(3789, func(agent core.Agent) {
+	core.AddWeaponEffect(3789, func(agent core.Agent, slot proto.ItemSlot) {
 		character := agent.GetCharacter()
-		mh := character.Equip[proto.ItemSlot_ItemSlotMainHand].Enchant.EffectID == 3789
-		oh := character.Equip[proto.ItemSlot_ItemSlotOffHand].Enchant.EffectID == 3789
-		if !mh && !oh {
-			return
-		}
-		procMask := core.GetMeleeProcMaskForHands(mh, oh)
+		procMask := core.GetMeleeProcMaskForHands(slot == proto.ItemSlot_ItemSlotMainHand, slot == proto.ItemSlot_ItemSlotOffHand)
 		ppmm := character.AutoAttacks.NewPPMManager(1.0, procMask)
 
 		// Modify only gear armor, including from agility
 		fivePercentOfArmor := (character.Equip.Stats()[stats.Armor] + 2.0*character.Equip.Stats()[stats.Agility]) * 0.05
-		procAuraMH := character.NewTemporaryStatsAura("Berserking MH Proc", core.ActionID{SpellID: 59620, Tag: 1}, stats.Stats{stats.AttackPower: 400, stats.RangedAttackPower: 400, stats.Armor: -fivePercentOfArmor}, time.Second*15)
-		procAuraOH := character.NewTemporaryStatsAura("Berserking OH Proc", core.ActionID{SpellID: 59620, Tag: 2}, stats.Stats{stats.AttackPower: 400, stats.RangedAttackPower: 400, stats.Armor: -fivePercentOfArmor}, time.Second*15)
+		name := core.Ternary(slot == proto.ItemSlot_ItemSlotMainHand, "MH", "OH")
+		tag := core.Ternary(slot == proto.ItemSlot_ItemSlotMainHand, 1, 2)
+		procAura := character.NewTemporaryStatsAura("Berserking "+name+" Proc", core.ActionID{SpellID: 59620, Tag: int32(tag)}, stats.Stats{stats.AttackPower: 400, stats.RangedAttackPower: 400, stats.Armor: -fivePercentOfArmor}, time.Second*15)
 
 		character.GetOrRegisterAura(core.Aura{
-			Label:    "Berserking (Enchant)",
+			Label:    name + " Berserking (Enchant)",
 			Duration: core.NeverExpires,
 			OnReset: func(aura *core.Aura, sim *core.Simulation) {
 				aura.Activate(sim)
 			},
 			OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
-				if !result.Landed() || !spell.ProcMask.Matches(core.ProcMaskMelee) {
+				hasEnchant := character.Equip[slot].Enchant.EffectID == 3789
+				if !hasEnchant || !result.Landed() || !spell.ProcMask.Matches(core.ProcMaskMelee) {
```

Comment:
This check works well enough for cases like this, but there's no way here to activate any ICDs, which will be needed for other effects.

The vague idea I had about this is that `AddWeaponEffect` could have a return value `type OnSwapItem(slot ItemSlot)` so that this code can provide a handler for item swaps. Then `core` would need to store that function somewhere and call it at the right time.